[WIP] Issue #18757 Tabbed compose preview

aldeka opened the pull request:
The preview vs write toggle button is now two tabs on the top right of
the compose area.

Known issues:
- It does add a little vertical height with the default Bootstrap styling. Could be modified.
- The colored stripe to the right of the stream name disappeared and I'm not sure why

<!-- What's this PR for?  (Just a link to an issue is fine.) -->
https://github.com/zulip/zulip/issues/18757

**Testing plan:** <!-- How have you tested? -->
- I tried with light and dark mode (are there other stylesheets?)
- TODO test with DMs
- TODO test with not-Chrome (haven't looked at what the browser SLA is yet)


**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
<img width="932" alt="Screen Shot 2021-06-29 at 00 03 53" src="https://user-images.githubusercontent.com/689247/123752361-918e7780-d86d-11eb-9bd7-01a08f18d542.png">

<img width="911" alt="Screen Shot 2021-06-29 at 00 04 17" src="https://user-images.githubusercontent.com/689247/123752384-94896800-d86d-11eb-95be-1f8bf83bc4cd.png">

<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

aldeka committed:
compose: Move draft preview functionality from eye button to tabs.

The ability to review a message draft in Zulip was not very
discoverable. Putting the preview viewer toggle above the textarea
makes it easier to find and clearer what will happen when you click
back and forth.

Fixes: #18757.

timabbott commented:
@aldeka welcome and thanks for working on this!  I took a quick look and here's my main feedback on the visuals:

* The backgrounds in the light theme seem more easy to parse than in the dark theme.
* On the left end of the image, it feels like there's a lot of extra lines, with the textbox border, tab divider line, and textarea border all stacked together:

  ![image](https://user-images.githubusercontent.com/2746074/123825252-9e35be80-d8b3-11eb-9a34-362b3b914d6d.png)

  There's a related somewhat awkward thing on both ends that the lines don't go all the way to the edge of the box, which feels a bit hard to parse:
  ![image](https://user-images.githubusercontent.com/2746074/123828268-303ec680-d8b6-11eb-9d80-efea03a8ea47.png)

  I notice that GitHub's similar widget addresses that design problem with their comment area Write/Preview widget by using a different background color for the compose area border region. 
  
  ![image](https://user-images.githubusercontent.com/2746074/123825894-1ef4ba80-d8b4-11eb-9e93-baf3d88a5f68.png)

  I'm not sure if that's the right answer, but we probably need to do something.

Another thing we'll want to address before merging is writing a commit message title consistent with the Zulip style.  You can see examples with `git log` or check out the Zulip commit message guidelines for more details: https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages

Also, if you end up changing the background color for the compose box, it's possible you'll want to do that as a second commit in this PR, either before or after the tabbed design change commit. It's very helpful for reviewing a series of code changes to have separate commits for separate ideas, but if you're not used to writing code that way, it's probably better to focus on getting the UI change where we want it and then figure out the Git commit structure for merging the code.  (One can always split commits with `git checkout -p` later).

ligmitz reviewed (commented):
The color stripe is set using `stream_bar.decorate()`. I guess there is some scope issues with the new change in the elements hierarchy.

aldeka committed:
compose: Add background color behind tabs.

This also adds some spacing modifications to improve alignment. With
the new coloration, the outer component no longer could provide the
padding so the subcomponents had be modified to provide equivalent
whitespace.

aldeka committed:
compose: Fix alignment of expand/collapse button and tabs.

aldeka committed:
compose: Ensure return to composition box on submit.

aldeka committed:
compose: Make compose/preview test more 'true-to-life'. [WIP]

The previous version was manually showing and hiding all the elements,
then checking their visibility. Better would be to simulate the UI
triggers that are supposed to cause this behavior.

Tried this as part of trying to figure why the tests were failing on my
PR, but didn't succeed.

aldeka commented:
I am currently stuck on the test failure for this PR.

The compose UI test `"enter_with_preview_open"` is failing consistently. `.visible()` is returning false for the compose textarea (and well as the preview widget!) and I have no idea why. The behavior in the real-life browser is as I would expect. I'm also not sure where the `.visible()` helper function is even coming from; I don't see it in the JQuery docs and `jquery-visible` isn't one of the package.json dependencies.
